src/markd/rules/table.cr

```diff
@@ -0,0 +1,116 @@
+module Markd::Rule
+  struct Table
+    include Rule
+
+    SEPARATOR_REGEX = /^(\|?\s*:{0,1}-:{0,1}+\s*)+(\||\s*)$/
+
+    def match(parser : Parser, container : Node) : MatchValue
+      if match?(parser) # Looks like the 1st line of a table
+        parser.close_unmatched_blocks
+        parser.add_child(Node::Type::Table, 0)
+
+        MatchValue::Leaf
+      else
+        MatchValue::None
+      end
+    end
+
+    def continue(parser : Parser, container : Node) : ContinueStatus
+      # Only continue if line looks like a divider or a table row
+      if match_continuation?(parser)
+        ContinueStatus::Continue
+      else
+        ContinueStatus::Stop
+      end
+    end
+
+    def token(parser : Parser, container : Node) : Nil
+      # The table contents are in container.text (except the leading | in each line)
+      # So, let's parse it and shove it into the tree
+
+      original_text = container.text.rstrip.split("\n").map do |l|
+        "#{l}"
+      end.join("\n")
+      lines = container.text.strip.split("\n")
+
+      row_sizes = lines[...2].map do |l|
+        l.strip.strip("|").split(/(?<!\\)\|/).size
+      end.uniq!
+
+      # Do we have a real table?
+      # * At least two lines
+      # * Second line is a divider
+      # * First two lines have the same number of cells
+
+      if lines.size < 2 || !lines[1].match(SEPARATOR_REGEX) ||
+         row_sizes.size != 1
+        # Not enough table or a broken table.
+        # We need to convert it into a paragraph
+        # I am fairly sure this is not supposed to work
+        container.type = Node::Type::Paragraph
+        # Patch the text to have the leading |s
+        container.text = original_text
+        return
+      end
+
+      max_row_size = row_sizes[0]
+      has_body = lines.size > 2
+      container.data["has_body"] = has_body
+
+      alignments = lines[1].strip.strip("|").split(/(?<!\\)\|/).map do |cell|
```

Comment:
nitpick: would it be useful to put this regex in a const as we're using it in several places?